adapters/msft/msft.go

```diff
@@ -0,0 +1,394 @@
+			errs = append(errs, fmt.Errorf("malformed impression ext for id '%s'", request.Imp[i].ID))
+			continue
+		}
+
+		if err := modifyImp(&request.Imp[i], impExt, displayManagerVerBuilder); err != nil {
+			errs = append(errs, fmt.Errorf("error building impression ext for id '%s'", request.Imp[i].ID))
+			continue
+		}
+
+		// ensure all impressions with member ids use the same member id
+		memberId := impExt.Bidder.Member
+		if memberId != 0 {
+			if uniqueMemberID == 0 {
+				uniqueMemberID = memberId
+			} else if uniqueMemberID != memberId {
+				errs = append(errs, errMemberIDMismatch)
+				return nil, errs
+			}
+		}
+
+		validImps = append(validImps, request.Imp[i])
+	}
+
+	if len(validImps) == 0 {
+		return nil, errs
+	}
+
+	request.Imp = validImps
+
+	requestURI := a.uri
+	if uniqueMemberID != 0 {
+		requestURI = appendMemberId(requestURI, uniqueMemberID)
+	}
+
+	if err := a.modifyRequestExt(request, requestInfo); err != nil {
+		return nil, append(errs, err)
+	}
+
+	requests, errors := buildRequests(request.Imp, request, requestURI.String())
+	return requests, append(errs, errors...)
+}
+
+func initDisplayManageVerBuilder(req *openrtb2.BidRequest) func() string {
+	if req.App == nil || len(req.App.Ext) == 0 {
+		return func() string { return "" }
+	}
+
+	var (
+		appExt json.RawMessage = req.App.Ext
+		result string
+		hasRun bool
+	)
+	return func() string {
+		if !hasRun {
+			source, err := jsonparser.GetString(appExt, openrtb_ext.PrebidExtKey, "source")
+			if err != nil {
+				hasRun = true
+				return result
+			}
+
+			version, err := jsonparser.GetString(appExt, openrtb_ext.PrebidExtKey, "version")
+			if err != nil {
+				hasRun = true
+				return result
+			}
+
+			result = fmt.Sprintf("%s-%s", source, version)
+			hasRun = true
+		}
+		return result
+	}
+}
+
+func modifyImp(imp *openrtb2.Imp, ext impExtIncoming, displayManagerVerBuilder func() string) error {
+	if ext.Bidder.InvCode != "" {
+		imp.TagID = ext.Bidder.InvCode
+	}
+
+	if imp.Banner != nil {
+		bannerCopy := *imp.Banner
+
+		if bannerCopy.W == nil && bannerCopy.H == nil && len(bannerCopy.Format) > 0 {
+			firstFormat := bannerCopy.Format[0]
+			bannerCopy.W = &(firstFormat.W)
+			bannerCopy.H = &(firstFormat.H)
+		}
+
+		if bannerCopy.API == nil {
+			bannerCopy.API = ext.Bidder.BannerFrameworks
+		}
+
+		imp.Banner = &bannerCopy
+	}
+
+	if len(imp.DisplayManagerVer) == 0 {
+		imp.DisplayManagerVer = displayManagerVerBuilder()
+	}
+
+	impExt := impExtOutgoing{
+		Appnexus: impExtOutgoingAppnexus{
+			PlacementID:       ext.Bidder.PlacementId,
+			AllowSmallerSizes: ext.Bidder.AllowSmallerSizes,
+			UsePmtRule:        ext.Bidder.UsePaymentRule,
+			Keywords:          ext.Bidder.Keywords,
+			TrafficSourceCode: ext.Bidder.TrafficSourceCode,
+			PubClick:          ext.Bidder.PubClick,
+			ExtInvCode:        ext.Bidder.ExtInvCode,
+			ExtImpID:          ext.Bidder.ExtImpId,
+		},
+		GPID: ext.GPID,
```

Comment:
Nitpick: Should `PlacementID` be renamed to `PlacementId`? Or was it, along with `PlacementID`, `UsePmtRule` and `ExtImpID` purposefully left different from `PlacementId`, `UsePaymentRule` and `ExtImpId`?